Popover: Refactor to render using React portals

aduth opened the pull request:
This pull request seeks to leverage React portals — a new (unstable) feature introduced in React 16 — to render the existing `Popover` component. The concept of portals has existed previously in the React ecosystem, through works such as [React-portal](https://github.com/tajo/react-portal) or [Calypso's `RootChild` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/root-child). The idea is that the rendering of the component occurs elsewhere in the DOM, usually at the root (body) element. This has a few benefits:

- Styles from parent nodes will not cascade into the popover
- It is immune from [phrasing content restrictions](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content) (e.g. `div` cannot be a descendant of `button`)

See included changes to existing Popover usage in PostVisibility as an example of the simplification permitted by the second of these benefits.

__Testing instructions:__

Verify that there are no regressions in the behavior of Popover controls, currently present in:

- Header inserter
- Post settings visibility
- Content inserter

Of note, the content inserter should demonstrate "bounds flipping" behavior. It should try to open upwards (on Demo content where scroll exists), else downward if space does not permit (on New post without any content).

Ensure that unit tests pass:

```
npm test
```

aduth committed:
Popover: Render as React portal

aduth committed:
Post Visibility: Remove unnecessary Popover wrapper

aduth committed:
Popover: Add README.md documentation

aduth committed:
Popover: Move outside click detection to Popover

Popover: Avoid close if click outside occurs in anchor parent

Popover: Stop propagation in visibility button popover

Amazingly, events still propagate from within a portaled node to its parent elements, likely because of React's synthetic events

Inserter: Close popover by click outside

Post Visibility: Close popover by click outside

aduth commented:
It was necessary to move click outside logic into Popover, which I'd intended to do eventually anyways, but was required to do so ahead of schedule only because otherwise the existing click outside handlers were assuming that the portaled content was "outside" and not allowing interaction within (e.g. selecting a block option from inserter menu). See c61b4bd.

youknowriad commented:
There's some highly engineered code in this PR :) 

I noticed a small regression tough in the header inserter when I hit the "insert" button a second time, the popover is not closed.

aduth commented:
@youknowriad Yeah, this has turned into a rabbit hole 😬  I noticed your issue as well and flipped to In Progress. There's a few interesting things going on here:

- Events on portaled nodes still bubble to ancestor elements, which is pretty amazing but also unexpected. In case of Post Visibility, the popover is inadvertently closed when clicking a visibility option because the click event is bubbling up to the button, which toggles itself closed. The solution is to stop propagation of the click event from reaching the button (63e21c6)
- With moving "click outside" logic into the popover itself, clicking on a toggle button is considered "outside" the popover, so we'd _want_ the popover to close, but then because we're clicking the toggle button it immediately reopens itself. The solution is to check, when closing by click outside, that the target of the click outside is not the button (b8099cf, ddae0d2)

aduth committed:
Inserter: Fix mobile popover height relative input

Icon button size was used as a stand-in for the input size. Use expected input size instead

aduth commented:
Noticing positioning was off on mobile, I started down another rabbit hole of fixing styling, noting an interesting gotcha that avoiding the style cascade is that assumptions about box sizing no longer hold true (i.e. need to apply `box-sizing` yourself to any portaled content). Which makes sense in retrospect, but was still unexpected.

The result of the work hopefully pays off. Compared to master, the height of the header inserter is more correct (5d2f030) and we restore the arrow, correctly positioned under the "+" icon:

`master`|`update/popover-portal`
---|---
![master](https://user-images.githubusercontent.com/1779930/29035767-d307836e-7b6a-11e7-89a6-081485a2db25.png)|![update/popover-portal](https://user-images.githubusercontent.com/1779930/29035768-d30995dc-7b6a-11e7-9c1c-fb7f352e2844.png)

cc @jasmussen for a style audit (and/or criticism on my desktop-first media query 😲 )

jasmussen commented:
Visually I think this looks great 👍👍!

> and/or criticism on my desktop-first media query

Can you elaborate? ;) Not sure what to look for.

aduth commented:
>Can you elaborate? ;) Not sure what to look for.

Just a [single selector using the inverse of the medium breakpoint](https://github.com/WordPress/gutenberg/blob/7db7f8e/editor/inserter/style.scss#L75-L83) since it'd be unrealistic to unset those properties in the reverse case.

jasmussen commented:
Ooooh dang, that's _controversial_! Our mixin is decidedly mobile first and here you go and _except to that rule 🤘

(It's fine, I kinda expected we'd need to do this at some point :) )

aduth commented:
@youknowriad Pulling out of commit comment, since it'll be lost on rebase:

>Should we do this consistently? I mean inside the popover component?

I see it being a common pain point, but at the same time, it should be the expected behavior. If I render a child and bind an event to the parent, I should expect that event occurring on the child to bubble to the parent.

It's in these specific use-cases cases in which we have button toggling mixed with click outside where it gets particularly sticky.

One uncertainty I have on second glance is if we might just apply event handlers to the Popover's anchor (span) node instead of the portal-ed node, since the same bubbling behavior will cause events to occur on the `span`.
